### ISSUE TITLE
Remote Relative Includes does not work on Windows?

### ISSUE THREAD

**Issue by deldrid1** [electricimp/Builder#66]
This is a followup to #42 and PR #64.  The issue is that "/" and "\" can be confused on Windows vs. Linux in remote includes (that may not match the target machine)

I think the offending line of code is this (or a lack of preprocessing that happens before this point). https://github.com/electricimp/Builder/blob/01f7443fa6ac49edfbed3bf5eaceeb508b70368d/src/Readers/GithubReader.js#L238

**Comment by betzrhodes** [comment]
Thanks looking into this.

**Comment by FedorUporov** [comment]
@deldrid1, could you please provide an example of include path, which does not work for you. I have some assumptions, but I want to study your case for deeper understanding of the issue.

**Comment by deldrid1** [comment]
This bug report may be a bit misleading...  There are some differences that I've noticed between Windows and macOS but in my repo this issue and #67 are very much intertwined.  Here's the very simple test cases I'm working with right now that don't expose Eaton's IP: https://github.com/deldrid1/Builder/tree/develop/spec/fixtures/sample-17.

It seems like there is some change in Builder's behavior with remote relative includes, even without the `--use-remote-relative-includes` flag based on my tests that are unclear in the documentation or my understanding.

Here are the examples I am using:
```bash
# 1) Expect to work when the repo is cloned and in the right working directory
pleasebuild spec/fixtures/sample-17/inc-a.nut 

# 2) Expect to work when the repo is cloned and in the right working directory
pleasebuild spec/fixtures/sample-17/inc-b.nut

# 3) Don't expect to work without --use-remote-relative-includes
pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-a.nut@develop

# 4) Don't expect to work without --use-remote-relative-includes
pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-b.nut@develop

# 5) Expect to work 
pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-a.nut@develop --use-remote-relative-includes

# 6) Desire to work per issue #67 
pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-b.nut@develop --use-remote-relative-includes
```

On macOS, 3 of the tests work the way that I would expect:
```bash
# macOS 10.14.4, node.js v10.15.3
# 1) PASS
# 2) PASS
# 3) is currently working, but shouldn't be???
# 4) PASS (the build fails, but possibly not for the right reason...)
# 5) FAIL - is NOT working
# 6) FAIL - is NOT working

AustinEldridge at Austins-MacBook-Pro in ~/GitRepos/Builder on develop [!?]
$ pleasebuild spec/fixtures/sample-17/inc-a.nut
// spec/fixtures/sample-17/lib/libFile.agent.nut

AustinEldridge at Austins-MacBook-Pro in ~/GitRepos/Builder on develop [!?]
$ pleasebuild spec/fixtures/sample-17/inc-b.nut
// spec/fixtures/sample-17/lib/libFile.agent.nut

AustinEldridge at Austins-MacBook-Pro in ~/GitRepos/Builder on develop [!?]
$ pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-a.nut@develop
// spec/fixtures/sample-17/lib/libFile.agent.nut

AustinEldridge at Austins-MacBook-Pro in ~/GitRepos/Builder on develop [!?]
$ pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-b.nut@develop
Local file "lib/libFile.agent.nut" not found (github:deldrid1/Builder/spec/fixtures/sample-17/inc-b.nut:1)

AustinEldridge at Austins-MacBook-Pro in ~/GitRepos/Builder on develop [!?]
$ pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-a.nut@develop --use-remote-relative-includes
Failed to get source "github:deldrid1/Builder/spec/fixtures/sample-17/spec/fixtures/sample-17/lib/libFile.agent.nut" from GitHub: Not Found (github:deldrid1/Builder/spec/fixtures/sample-17/inc-a.nut:1)

AustinEldridge at Austins-MacBook-Pro in ~/GitRepos/Builder on develop [!?]
$ pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-b.nut@develop --use-remote-relative-includes
Failed to get source "github:deldrid1/Builder/spec/fixtures/sample-17/lib/libFile.agent.nut" from GitHub: Not Found (github:deldrid1/Builder/spec/fixtures/sample-17/inc-b.nut:1)
```

On Windows 10, 4 of the tests work the way that I would expect.  There is also some weirdness with the path separators in the logs that you can see.  
```bash
# Windows 10 Pro Version 1709 (OS Build 16299.1029), node.js v9.10.1
# 1) PASS
# 2) PASS
# 3) is currently working, but shouldn't be???
# 4) PASS (the build fails, but possibly not for the right reason...)
# 5) PASS 
# 6) FAIL - is NOT working

C:\Users\E0088611\Desktop\Builder>pleasebuild spec/fixtures/sample-17/inc-a.nut
// spec/fixtures/sample-17/lib/libFile.agent.nut

C:\Users\E0088611\Desktop\Builder>pleasebuild spec/fixtures/sample-17/inc-b.nut
// spec/fixtures/sample-17/lib/libFile.agent.nut

C:\Users\E0088611\Desktop\Builder>pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-a.nut@develop
// spec/fixtures/sample-17/lib/libFile.agent.nut

C:\Users\E0088611\Desktop\Builder>pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-b.nut@develop
Local file "lib\libFile.agent.nut" not found (github:deldrid1\Builder\spec\fixtures\sample-17\inc-b.nut:1)

C:\Users\E0088611\Desktop\Builder>pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-a.nut@develop --use-remote-relative-includes
// spec/fixtures/sample-17/lib/libFile.agent.nut

C:\Users\E0088611\Desktop\Builder>pleasebuild github:deldrid1/Builder/spec/fixtures/sample-17/inc-b.nut@develop --use-remote-relative-includes
Local file "lib\libFile.agent.nut" not found (github:deldrid1\Builder\spec\fixtures\sample-17\inc-b.nut:1)
```

**Comment by ppetrosh** [comment]
Hi, @deldrid1 I'm sorry for overseeing these issues on Windows. We are looking into them to make sure we do the right fix and also adjust the process so that we don't miss things like that in the future. Can you please help us prioritize it so that we unblock your process. Which of these issues is critical for you to get fixed ASAP? For the rest, I would take some time to do them right and thoroughly retest across all the platforms.

**Comment by deldrid1** [comment]
It's an easy thing to miss - I did until we started beginning to try to use the tool on our production repos...

Unfortunately, our code is old, with many (inexperienced) hands contributing and therefore it has a lot of inconsistencies.  My preference is to get example ```6)``` working, as this is the most readable code.

I'm not sure how to prioritize the work - ultimately, I need cross-platform way to do remote relative includes...  I'll follow-up with an email and a more concrete example.

**Comment by betzrhodes** [comment]
macOS 10.14.4 node v11.3.0
I've cloned your repo and pointed it to a repo in my Github account. I've run the same tests as you have and I'm getting slightly different results. In my local repo (the one on my computer) I added some logs - this makes it very clear if local files vs remote files are being used, and also what file and path builder is referencing. I have not added or committed the updated files to Github, and here are the result I got when running the same commands you have above. My results: 

1 - File builds, same results as yours
```
Elizabeths-MacBook-Pro:Builder electricimp$ pleasebuild spec/fixtures/sample-17/inc-a.nut
// "FILE: " inc-a.nut
// -----------------------------------------------
// "FILE: " libFile.agent.nut
// spec/fixtures/sample-17/lib/libFile.agent.nut
// "PATH: " spec/fixtures/sample-17/lib
// -----------------------------------------------
// "PATH: " spec/fixtures/sample-17
// Include is absolute
```

2 - File builds, same results as yours
```
Elizabeths-MacBook-Pro:Builder electricimp$ pleasebuild spec/fixtures/sample-17/inc-b.nut
// "FILE: " inc-b.nut
// -----------------------------------------------
// "FILE: " libFile.agent.nut
// spec/fixtures/sample-17/lib/libFile.agent.nut
// "PATH: " lib
// -----------------------------------------------
// "PATH: " spec/fixtures/sample-17
// Include is not absolute
```

3 - File builds, same results as yours
```
Elizabeths-MacBook-Pro:Builder electricimp$ pleasebuild github:betzrhodes/BuilderRelativeIncludeTesting/spec/fixtures/sample-17/inc-a.nut@develop
// -----------------------------------------------
// "FILE: " libFile.agent.nut
// spec/fixtures/sample-17/lib/libFile.agent.nut
// "PATH: " spec/fixtures/sample-17/lib
// -----------------------------------------------
```

4 - Build Fails, same results as yours
```
Elizabeths-MacBook-Pro:Builder electricimp$ pleasebuild github:betzrhodes/BuilderRelativeIncludeTesting/spec/fixtures/sample-17/inc-b.nut@develop
Local file "lib/libFile.agent.nut" not found (github:betzrhodes/BuilderRelativeIncludeTesting/spec/fixtures/sample-17/inc-b.nut:1)
```

5 - Build Fails, same result as yours
```
Elizabeths-MacBook-Pro:Builder electricimp$ pleasebuild github:betzrhodes/BuilderRelativeIncludeTesting/spec/fixtures/sample-17/inc-a.nut@develop --use-remote-relative-includes
Failed to get source "github:betzrhodes/BuilderRelativeIncludeTesting/spec/fixtures/sample-17/spec/fixtures/sample-17/lib/libFile.agent.nut" from GitHub: Not Found (github:betzrhodes/BuilderRelativeIncludeTesting/spec/fixtures/sample-17/inc-a.nut:1)
```

6 - File builds, different results as yours
```
Elizabeths-MacBook-Pro:Builder electricimp$ pleasebuild github:betzrhodes/BuilderRelativeIncludeTesting/spec/fixtures/sample-17/inc-b.nut@develop --use-remote-relative-includes
// spec/fixtures/sample-17/lib/libFile.agent.nut
```

**Comment by betzrhodes** [comment]
The repo I was using is public and should be a clone of yours - https://github.com/betzrhodes/BuilderRelativeIncludeTesting/tree/develop/spec/fixtures/sample-17

**Comment by betzrhodes** [comment]
My updated local files
 
libFile.agent.nut:
```
// -----------------------------------------------
// "FILE: " @{__FILE__}
// spec/fixtures/sample-17/lib/libFile.agent.nut
// "PATH: " @{__PATH__}
// -----------------------------------------------
```

inc-a.nut: 
```
// "FILE: " @{__FILE__}
@include once "spec/fixtures/sample-17/lib/libFile.agent.nut"
// "PATH: " @{__PATH__}
// Include is absolute
```

inc-b.nut
```
// "FILE: " @{__FILE__}
@include once "./lib/libFile.agent.nut"
// "PATH: " @{__PATH__}
// Include is not absolute
```

**Comment by betzrhodes** [comment]
Windows issues have been addressed in release v2.8.1 which is now available. I'd like to track include issues in a different thread - we can use #67 for now.
